gusto/equations.py

```diff
@@ -276,13 +277,20 @@ def generate_linear_terms(self, residual, terms_to_linearise):
         # TODO: Neaten up the `terms_to_linearise` variable. This should not be
         # a dictionary, it should be a filter of some sort
 
-        # Loop through prognostic variables and linearise any specified terms
-        for field in self.field_names:
-            residual = residual.label_map(
-                # Extract all terms to be linearised for this field's equation that haven't already been
-                lambda t: (not t.has_label(linearisation) and (t.get(prognostic) == field)
-                           and any(t.has_label(*terms_to_linearise[field], return_tuple=True))),
-                lambda t: linearisation(t, LabelledForm(t).label_map(all_terms, replace_subject(self.trials)).terms[0]))
+        # This is quite in-depth so it is clearer to loop through the terms in
+        # the residual rather than to use the label_map
+        for term_idx, t in enumerate(residual.terms):
+            field = t.get(prognostic)
+            if (not t.has_label(linearisation)
+                and any(t.has_label(*terms_to_linearise[field], return_tuple=True))):
+                # Linear form is obtained by taking derivative of original term
+                linear_form = action(ufl.derivative(t.form, self.X), self.trials)
+                # Make term by adding labels from original term to form
+                raw_linear_term = Term(linear_form, t.labels)
+                # Replace the prognostic variables with their reference state
+                linear_term = replace_subject(self.X_ref)(raw_linear_term)
+                # Add the linearisation to the original term through a label
+                residual.terms[term_idx] = linearisation(t, linear_term)
```

Comment:
Why can't you use `label_map` any more?

Reply:
Ah, I see this is deliberate

Reply:
```python
from functools import partial

def should_linearise(term, field):
    return (
        not term.has_label(linearisation)
        and term.get(prognostic) == field
        and any(term.has_label(*terms_to_linearise[field], return_tuple=True))
   )


def linearise(term, X, X_ref, du):
    linear_term = Term(action(ufl.derivative(term.form, X), du), term.labels)
    linear_term = replace_subject(X_ref, linear_term)
    return linearisation(term, linear_term)


for field in self.field_name:
    residual = residual.label_map(
        partial(should_linearise, field=field),
        partial(linearise, self.X, self.X_ref, self.trials)
    )
```

??

Reply:
Thanks, that looks great! I'll lift that exactly as is.

I like the `keep` suggestion too.

Reply:
(Fully untested btw)

Reply:
It all works for me, just needed to change `self.field_name` to `self.field_names` (which highlights that it is confusing to have both of these variables!)